### PR TITLE
New _find_starts_of_groups()

### PR DESCRIPTION
**Pull request by xluo127**
Hi, would you like to have a look at this? The new function is nearly twice as fast as the old one.

I have carefully checked everything for `_find_starts_of_groups`. There is no change for input, output, and syntax. 

The only thing is that I have to remove two arguments in the in-place function `_find_starts_of_groups!` to get better performance.  In #17 you said that you do not want to change the syntax of this function, too. Maybe you plan to add something new in `_find_starts_of_groups!`  in the future?

I am just proposing the new code to get better performance. If it is not appropriate, you can just ignore the commits.

### CODE REVIEW

**Comment by sl-solution** [comment]
Thanks for the contribution,
would you please apply the following changes:
 
* provide a set of benchmarks to support your PR
* replace range_cpy with inbits = zeros(Bool, nrow(ds))
* change `format` to `f` inside the function (because it works for any f), i.e. in this function call`@inbounds ranges_cpy[j] = ranges_cpy[j]==1 ? 1 : !isequal(format(x[j]), format(x[j-1]))`
* add a series of tests for the new function

**Comment by xluo127** [comment]
OK, I will manage to do so.

**Comment by xluo127** [comment]
Hi.

I have updated the code so that it can be much faster if `colsidx <= index(ds).sortedcols`.

`f0` in the code below is the current old version of `_find_starts_of_groups`.

```
julia> ds = Dataset(rand(1:100,1000000,8), :auto);

julia> groupby!(ds, 1:3)
1000000×8 Grouped Dataset with 632574 groups
Grouped by: x1, x2, x3

julia> @btime f0(ds, colsidx, Val{Int32}()) # The old version.
  14.990 ms (16 allocations: 7.63 MiB)
([1, 2, 3, 4], Int32[1, 2, 3, 4, 5, 6, 7, 8, 9, 10  …  0, 0, 0, 0, 0, 0, 0, 0, 0, 0], 996418)

julia> @btime IMD._find_starts_of_groups(ds, colsidx, Val{Int32}()) # The new version in the PR.
  5.310 ms (97 allocations: 4.78 MiB)
([1, 2, 3, 4], Int32[1, 2, 3, 4, 5, 6, 7, 8, 9, 10  …  0, 0, 0, 0, 0, 0, 0, 0, 0, 0], 996418)

julia> ds1 = Dataset(rand(1:10, 10^7,5), :auto);

julia> groupby!(ds1, :);

julia> ds1
10000000×5 Grouped Dataset with 100000 groups
Grouped by: x1, x2, x3, x4, x5

julia> @btime f0(ds1, colsidx, Val{Int32}()) # The old version.
  84.481 ms (14 allocations: 76.29 MiB)
([1, 2, 3, 4], Int32[1, 1021, 2003, 3077, 4048, 5096, 6075, 7059, 8115, 9123  …  0, 0, 0, 0, 0, 0, 0, 0, 0, 0], 10000)

julia> @btime IMD._find_starts_of_groups(ds1, colsidx, Val{Int32}()) # The new version in the PR.
  17.930 ms (97 allocations: 47.69 MiB)
([1, 2, 3, 4], Int32[1, 1021, 2003, 3077, 4048, 5096, 6075, 7059, 8115, 9123  …  0, 0, 0, 0, 0, 0, 0, 0, 0, 0], 10000)

```

I am not sure where should I put the tests for this `find_starts_of_groups`. And I would like to add some tests for your sorting & grouping functions, do you have some suggestions?

**Comment by sl-solution** [comment]
Thanks for you contribution,

* no need to use `IMD` identifier, e.g. you can use` index(ds)` directly
* `ds` can be a` Dataset` or a` GroupBy` and you should use functions defined at the bottom of src/sort/groupby.jl to access different attributes of `ds`
* the sorting columns are not needed to be consecutive columns for example they can be `sort(ds, [1, 4, 2, 10])`, thus `colsidx <= index(ds).sortedcols` may not be correct
* use `true` and `false` for Bool vectors, rather than `1` and `0`
* put your tests as a new set in test/sort.jl


For grouping and test you can create  test sets inside the files which are created for this purpose. Usually, your tests should contain different general use cases of the functions and any corner cases that you can come up with. You can look or use test sets from other packages (DataFrames.jl, pandas, ...)

**Comment by sl-solution** [comment]
You can find more information about grouping and sorting in the package manual.

**Comment by xluo127** [comment]
Sorry that I forget to remove the `IMD`, I use it just for measuring performance. 

I will carefully check these points.

**Comment by xluo127** [comment]
I have updated the code based on your comments.

Some common cases of using `_find_starts_of_groups` are added in the test file. 

I will open another PR for tests of grouping and sorting later.

File: src/other/utils.jl [hunk 1]
```diff
@@ -479,48 +479,38 @@ function _gather_groups_old_version(ds, cols, ::Val{T}; mapformats = false) wher
...
-                ranges[cnt] = i + 1
-                cnt += 1
-            end
-        end
-    end
-    @inbounds for j in 1:(cnt - 1)
-        ranges_cpy[j] = ranges[j]
+function _find_starts_of_groups!(x, f, inbits)
+    Threads.@threads for j in 2:length(inbits)
+        @inbounds inbits[j] = inbits[j]==true ? true : !isequal(f(x[j]), f(x[j-1]))
```
Comment (sl-solution) [inline src/other/utils.jl:512]: why not `inbits[j] ? nothing : inbits[j] = !isequal(f(x[j]), f(x[j-1]))`

File: src/other/utils.jl [hunk 2]
```diff
@@ -479,48 +479,38 @@ function _gather_groups_old_version(ds, cols, ::Val{T}; mapformats = false) wher
 end
 
 # ds assumes is grouped based on cols and groups are gathered togther
-function  _find_starts_of_groups(ds, cols::Vector, ::Val{T}; mapformats = true) where T
+function _find_starts_of_groups(ds, cols::Vector, ::Val{T}; mapformats = true) where T
```
Comment (sl-solution) [inline src/other/utils.jl:482]: finding groups from  the last column was a good idea, why you dropped it?

File: test/sort.jl [hunk 3]
```diff
@@ -142,3 +142,46 @@ using InMemoryDatasets, PooledArrays, Random, Test, CategoricalArrays
     @test sortperm(dsl, :)==sortperm(dsr, :)
     @test sortperm(dsl, :, rev=true)==sortperm(dsr, :, rev=true)
 end
+
+@testset "_find_starts_of_groups" begin
```
Comment (sl-solution) [inline test/sort.jl:146]: You need to add some tests where you have different type of data (and combination of them) such as PooledArray, CategoricalArray, Float64,...
consider cases where there are missing values in data
consider cases where you have a larger data set perhaps 10,000 observations with 10 columns/ with many levels/ with just a few levels

File: src/other/utils.jl (hunk 2 above)
Comment (xluo127) [reply src/other/utils.jl:482]: I think misunderstand your words before.
```
julia> [1, 3] <= 1:3
false
```
Maybe you mean in this case using `colsidx <= index(ds).sortedcols` is not correct? I can add it back and change this line.

File: src/other/utils.jl (hunk 1 above)
Comment (xluo127) [reply src/other/utils.jl:512]: `true` means the start of a group, `inbits[j]==true` means for at least one previous column, `f(x[j]) != f(x[j-1])`, so no need to do comparison for this `j`.

If we initialize the `inbits` using `zeros(Bool, nrow(ds))`, why there is `nothing` in the vector?

File: test/sort.jl (hunk 3 above)
Comment (xluo127) [reply test/sort.jl:146]: OK, I will add more test cases.

File: src/other/utils.jl (hunk 1 above)
Comment (sl-solution) [reply src/other/utils.jl:512]: `inbits[j] ? nothing : inbits[j] = !isequal(f(x[j]), f(x[j-1]))` actually translate to 
```julia
if !inbits[j]
     inbits[j] = !isequal(f(x[j]), f(x[j-1]))
end
```
no need to worry about `nothing`

File: src/other/utils.jl (hunk 2 above)
Comment (sl-solution) [reply src/other/utils.jl:482]: This needs more care, you may refer to `_check_for_fast_sort` in `src/sort/sortperm.jl`, or use this function for your check.

File: src/other/utils.jl (hunk 1 above)
Comment (xluo127) [reply src/other/utils.jl:512]: OK. 
When we start from `zeros(Bool, nrow(ds))`, all positions except the first row should be `false`. If `inbits[j]` is `true`, we will never change that because at least one column `xi` exists `f(xi[j]) != f(xi[j-1])`. This `j`-th row will be the start of group so no need to do `!isequal(f(x[j]), f(x[j-1]))`.

Things will be different here if grouping information of the current `ds` can be used. I will carefully check this case.

File: src/other/utils.jl (hunk 2 above)
Comment (xluo127) [reply src/other/utils.jl:482]: It seems that `_find_starts_of_groups` is only used when `_check_for_ast_sort` returns 1 in `src/sort/sortperm.jl`. In this case, `colsidx` should be a subset of `sortedcols`.

Do you mean that I can change this `colsidx <= index(ds).sortedcols` to something like 
```
if length(colsidx) < length(scols)
        if colsidx == view(scols, 1:length(colsidx)) && rev == view(revs, 1:length(colsidx)) && mapformats == fmt
```
in  `_check_for_ast_sort`?

In this case, I can use `starts` and `ngroups[]` from the current `ds` to get better performance.

Comment (sl-solution) [reply src/other/utils.jl:482]: something like that
Note that `rev` may not be needed in your function but checking if the sorted data set used format for sorting is important.
also note that you should be aware of the data set passed to your function, it can be either Dataset, or GroupBy (special view of a data set)

Comment (xluo127) [reply src/other/utils.jl:482]: Yes. As you told me, I will use functions in src/sort/groupby.jl to extract information so that it works for both `Dataset` and `GroupBy`.

**Comment by xluo127** [comment]
It seems that `getformat` does not work on `GroupBy`, I currently use `typeof(ds) <: AbstractDataset` to skip `getformat` on `GroupBy`.

**Comment by sl-solution** [comment]
you cannot skip getting format information, since if `mapformats=true` the formatted values should be used.

**Comment by xluo127** [comment]
`getformat(ds, cols)` does not work if `ds` is a `GroupBy`. If there anywhere I can find the format of it?

**Comment by sl-solution** [comment]
I added `getformat` for `GroupBy`, it should work now

**Comment by xluo127** [comment]
Hi, I have updated the code and also the tests.

**Comment by xluo127** [comment]
There is a problem with the function. It can now work for `GroupBy` and `Dataset`, but not for `SubDataset`. The reason is that functions like `_group_starts` in src/sort/groupby.jl does not work on `SubDataset`. I now skip the `SubDataset` type in the function to pass the tests. Otherwise, there will be errors in line 83 of test/sort.jl.

I now initialize the bool vector using `zeros(Bool, nrow(ds))` if `ds` is a `SubDataset`. It is not perfect, but it works.

**Comment by sl-solution** [comment]
Few points:

 * there is no `starts`, `ngroups` attributes for `SubDataset`, so it is ok to fallback to default function in this case.
 * you can not use `starts`,` ngroups` just by checking sorted columns, e.g. what happen if the data set is sorted or grouped by `mapformats = false` or `mapformats = true`?
 * Another issue is that a `GroupBy` data set is a view of its parent but not the same permutation as its parent. the `starts` are based on the permuted column not the parent permutation.
 * You may fix these issues and add tests to cover them.

**Comment by xluo127** [comment]
For the second point, I can add `mapformats == _get_fmt(ds)` to check it.
For the third point, do you mean you want results based on parent of a `GroupBy` rather than itself?
```
julia> ds = Dataset(x1= [1,1,1 ], x2=[1,2,1])
3×2 Dataset
 Row │ x1        x2
     │ identity  identity
     │ Int64?    Int64?
─────┼────────────────────
   1 │        1         1
   2 │        1         2
   3 │        1         1

julia> gb = groupby(ds, 2)
3×2 View of Grouped Dataset, Grouped by: x2
 x1        x2
 identity  identity
 Int64?    Int64?
────────────────────
        1         1
        1         1
        1         2
```
If we pass this `gb` & `cols=2` to `_find_starts_of_groups`, now starts will be `[1, 3]`. Do you mean I have to let it works on its parent, so that the result is `[1, 2, 3]`?

**Comment by xluo127** [comment]
Maybe you mean the problem is here:
```
julia> j = 2
2

julia> IMD._columns(gb)[j]
3-element Vector{Union{Missing, Int64}}:
 1
 2
 1

julia> IMD._columns(ds)[j]
3-element Vector{Union{Missing, Int64}}:
 1
 2
 1
```
The `_columns` will use the permutation of its parent which is incorrect. Maybe `_columns(gb)[j][gb.perm]` is the correct way of doing this?
```
julia> IMD._columns(gb)[j][gb.perm]
3-element Vector{Union{Missing, Int64}}:
 1
 1
 2
```

**Comment by sl-solution** [comment]
Yes,
however, note that`IMD._columns(gb)[j][gb.perm]`allocates memory, you must use views to avoid memory allocation. 
Also note that _get_perms() returns the permutation of a GroupBy

**Comment by xluo127** [comment]
OK, I see!

**Comment by xluo127** [comment]
Hi, I have updated the code and the tests, the function now should work for the cases you mentioned above.

**Comment by sl-solution** [comment]
I added some changes, if you are happy I can merge it. (I remove the fast past)

**Comment by xluo127** [comment]
Sure! The faster version works but is not readable. You may have to add additional `if` to let the fast code work when you update something new in the future, which is not convenient for you. You can merge it.
